#! /bin/sh -e

# DP: Turn on -Wl,-z,relro by default.

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p0 < $0
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0
Index: gcc/gcc.c
===================================================================
--- gcc.orig/gcc.c	2008-05-01 13:41:06.000000000 -0700
+++ gcc/gcc.c	2008-05-01 13:41:09.000000000 -0700
@@ -726,6 +726,7 @@
 #define LINK_COMMAND_SPEC "\
 %{!fsyntax-only:%{!c:%{!M:%{!MM:%{!E:%{!S:\
     %(linker) %l " LINK_PIE_SPEC "%X %{o*} %{A} %{d} %{e*} %{m} %{N} %{n} %{r}\
+    -z relro\
     %{s} %{t} %{u*} %{x} %{z} %{Z} %{!A:%{!nostdlib:%{!nostartfiles:%S}}}\
     %{static:} %{L*} %(mfwrap) %(link_libgcc) %o\
     %{fopenmp|ftree-parallelize-loops=*:%:include(libgomp.spec)%(link_gomp)} %(mflib)\
Index: gcc/doc/invoke.texi
===================================================================
--- gcc.orig/doc/invoke.texi	2008-05-01 13:42:46.000000000 -0700
+++ gcc/doc/invoke.texi	2008-05-01 13:45:04.000000000 -0700
@@ -7581,6 +7581,9 @@
 Pass @var{option} as an option to the linker.  If @var{option} contains
 commas, it is split into multiple options at the commas.
 
+NOTE: In Ubuntu 8.10 and later versions, for LDFLAGS, the
+@option{-Wl,-z,relro} is used.  To disable, use @option{-Wl,-z,norelro}.
+
 @item -u @var{symbol}
 @opindex u
 Pretend the symbol @var{symbol} is undefined, to force linking of
